// Inserting one single account
Account acc = new Account(Name='CLR Infotech', Phone='9988998899');
acc.Rating = 'Hot';
insert acc;


List<Account> accounts = new List<Account>();
Account acc1 = new Account(Name='Test Account 1');
Account acc2 = new Account(Name='Test Account 2');
Account acc3 = new Account(Name='Test Account 3');
Account acc4 = new Account(Name='Test Account 4');
Account acc5 = new Account(Name='Test Account 5');
accounts.add(acc1);
accounts.add(acc2);
accounts.add(acc3);
accounts.add(acc4);
accounts.add(acc5);
// Inserting a list of account
insert accounts;

// Alternate way to insert the accounts
Database.insert(accounts, false);

// Inserting records one by one
// Each record is inserted with separate insert statement
// Debug logs
// Number of DML statements: 5 out of 150
// Number of DML rows: 5 out of 10000
Account acc1 = new Account(Name='Test Account 1');
insert acc1;
Account acc2 = new Account(Name='Test Account 2');
insert acc2;
Account acc3 = new Account(Name='Test Account 3');
insert acc3;
Account acc4 = new Account(Name='Test Account 4');
insert acc4;
Account acc5 = new Account(Name='Test Account 5');
insert acc5;

//.............................Exception......................

// Math Exception
Double result = 10/0;
System.debug('Result: '+result);

// DML Exception
Account accRec = new Account();
insert accRec;


// Null Pointer Exception
String str;
System.debug(str.toLowerCase());


// throw an exception Programmatically
throw new NullPointerException();


//................................Try-catch-block....................

System.debug('Before Exception');
try{
    // Math Exception
    Double result = 10/0;
    System.debug('Result: '+result);

    // DML Exception
    // Account accRec = new Account();
    // insert accRec;
    System.debug('Empty try block ');
} catch(Exception e){
    System.debug('Caught generic exception ');
} finally {
    System.debug('finally called ');
}
System.debug('After Exception');


//.........................Exception Methods............................

System.debug('Before Exception');
try{
    // Math Exception
    Double result = 10/0;
    System.debug('Result: '+result);

    // DML Exception
    // Account accRec = new Account();
    // insert accRec;
    System.debug('Empty try block ');
} catch(Exception e){
    system.debug('Cause '+e.getCause());
    system.debug('Message '+e.getMessage());
    system.debug('Line Number '+e.getLineNumber());
    system.debug('Stack Trace '+e.getStackTraceString ());
} finally {
    System.debug('finally called ');
}
System.debug('After Exception');


//.............................throw a custom Exception....................

System.debug('Before Exception');
try{
   throw new ProcessException('My custom exception');
} catch(Exception e){
    system.debug('Cause '+e.getCause());
    system.debug('Message '+e.getMessage());
    system.debug('Line Number '+e.getLineNumber());
    system.debug('Stack Trace '+e.getStackTraceString ());
} finally {
    System.debug('finally called ');
}
System.debug('After Exception');